### PR TITLE
Miscellaneous tests to add (from legacy notebooks) 

### PR DESCRIPTION
**Pull request by richarddushime**
This PR  fixes a small part of #16 

3 notebooks 
2_polblogs_benchmark.ipynb:
3_verify_shapes.ipynb:
5_sampling.ipynb:

### CODE REVIEW

File: tests/test_manifolds.py [hunk 1]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+            for M in pm.P:
+                end = start + M.ambient_dim
+                slices.append(slice(start, end))
+                start = end
+                
+            pdist2_sum = sum(M.pdist2(X1[:, slc]) for M, slc in zip(pm.P, slices))
+            dist2_sum = sum(M.dist2(X1[:, slc], X2[:, slc]) for M, slc in zip(pm.P, slices))
+            
+            assert torch.allclose(pdist2_total, pdist2_sum, atol=1e-5), "pdist2 does not match sum of component pdist2"
+            assert torch.allclose(dist2_total, dist2_sum, atol=1e-5), "dist2 does not match sum of component dist2"
```
Comment (pchlenski) [inline tests/test_manifolds.py]: These tests would error as written because the single manifold class `Manifold` does not have the `P` attribute. This should be moved to a separate function, or possibly into `test_product_manifold_methods` instead.

File: tests/test_manifolds.py [hunk 2]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
 
         _shared_tests(pm, X1, X2, is_euclidean=all(M.curvature == 0 for M in pm.P))
-
+        # dist2 and pdist2 are the sum of component dist2 and pdist2
+        if len(pm.P) > 1:
+            pdist2_total = pm.pdist2(X1)
+            dist2_total = pm.dist2(X1, X2)
+            
+            # Compute dimension slices manually
+            slices = []
```
Comment (pchlenski) [inline tests/test_manifolds.py]: Please use the purpose-built `pm.factorize` method to slice the manifold instead

File: tests/test_manifolds.py [hunk 3]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+            assert torch.allclose(pdist2_total, pdist2_sum, atol=1e-5), "pdist2 does not match sum of component pdist2"
+            assert torch.allclose(dist2_total, dist2_sum, atol=1e-5), "dist2 does not match sum of component dist2"
         # Also test gaussian mixture
         X, y = pm.gaussian_mixture(num_points=100, num_classes=2, seed=42, adjust_for_dims=True)
+
+        # Test that dist_component_by_manifold contributions sum to 1
+        from manify.embedders._losses import dist_component_by_manifold # type: ignore
+        if len(pm.P) > 1:  # Requires multiple components
+            contributions = dist_component_by_manifold(pm, X1)
+            assert torch.isclose(torch.tensor(sum(contributions)), torch.tensor(1.0), atol=1e-5), "Contributions do not sum to 1"
```
Comment (pchlenski) [inline tests/test_manifolds.py]: Similar to [this comment above](https://github.com/pchlenski/manify/pull/22/files#r2252896141), this code will error when run with the `Manifold` class; please move this as well

File: tests/test_manifolds.py [hunk 4]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+        mu = m.sample(z_mean=m.mu0)
+        Sigma = torch.diag(torch.randn(m.dim)) ** 2
+        
+        # Sample points from wrapped normal distribution
+        samples = m.sample(n_samples=N_SAMPLES, z_mean=mu, sigma=Sigma)
+        
+        # Compute distances to origin
+        distances_to_origin = m.dist(samples, m.mu0.unsqueeze(0)).squeeze()
+        
+        # For hyperbolic and Euclidean manifolds, distances should be consistent
```
Comment (pchlenski) [inline tests/test_manifolds.py]: Perhaps we can have something like this instead:

```python
max_dist = float('inf') if K > 0 else torch.pi / torch.sqrt(K)

# Rest of the code does not have if K <= 0 / else control flow any longer:

# Check that distances are reasonable (not all zero, not all infinite)
assert distances_to_origin.min() > 0, f"Distances to origin should be positive for K={K}"
assert distances_to_origin.max() < max_dist, f"Distances to origin should be less than {max_dist} for K={K}"

# Check that distances have reasonable variance (not all identical)
assert distances_to_origin.std() > 1e-6, f"Distances to origin should have variance for K={K}"
```

File: tests/test_manifolds.py [hunk 5]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+            assert distances_to_origin.max() < float('inf'), f"Distances to origin should be finite for K={K}"
+            
+            # Check that distances have reasonable variance (not all identical)
+            assert distances_to_origin.std() > 1e-6, f"Distances to origin should have variance for K={K}"
+        
+        # For spherical manifolds with high curvature, distances might be more constrained
+        elif K > 0:
+            # Spherical manifolds have bounded distances
+            max_possible_distance = math.pi / math.sqrt(K)
+            assert distances_to_origin.max() <= max_possible_distance, f"Distances should be bounded for spherical K={K}"
```
Comment (pchlenski) [inline tests/test_manifolds.py]: In general, we have much stronger expectations here: that the distance to origin should be, on average, close to the distance of `mu` to the origin, and that the standard deviation of the distance should be related to `Sigma` in some way (need to figure out what the exact relation is)

File: tests/test_manifolds.py [hunk 6]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+
+
+def test_log_likelihood_differences():
```
Comment (pchlenski) [inline tests/test_manifolds.py]: It seems like `test_log_likelihood_differences` and `test_product_manifold_log_likelihood_differences` can be added to the shared manifold tests instead, letting the loop over curvatures in `test_manifold_methods` and over signatures in `def test_product_manifold_methods` handle this variation.

File: tests/test_manifolds.py [hunk 7]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+        
+        # Check that the difference is reasonable (not too negative)
+        # Allow for some numerical tolerance - log-likelihood differences can be negative
+        assert log_likelihood_diff.mean() > -500.0, f"Log-likelihood difference should not be too negative for signature={signature}"
+        
+        # Check that individual differences are reasonable
+        assert log_likelihood_diff.std() > 0, f"Log-likelihood differences should have variance for signature={signature}"
+
+
+def test_kl_divergence_equivalence():
```
Comment (pchlenski) [inline tests/test_manifolds.py]: Same as this [comment above](https://github.com/pchlenski/manify/pull/22/files#r2252913500), please incorporate this into shared tests rather than writing separate loops over curvatures/signatures

Comment (pchlenski) [reply tests/test_manifolds.py]: "This" meaning KL divergence equivalence tests for single/product manifolds

File: tests/test_manifolds.py [hunk 8]
```diff
@@ -172,6 +181,330 @@ def test_product_manifold_methods():
...
+        assert torch.isfinite(kl_divergence_approx), f"KL divergence should be finite for signature={signature}"
+
+
+def test_sampling_consistency():
+    """Test that sampling produces consistent results with different input formats."""
+    print("Testing sampling consistency...")
+    
+    curvatures = [-1.0, 0.0, 1.0]
+    
+    for K in curvatures:
```
Comment (pchlenski) [inline tests/test_manifolds.py:313]: This could be strengthened by setting random seeds specifically: in that case, we should be able to verify that setting n_samples or using stacked z_mean are equivalent, as e.g. in this code from my original `_shared_tests` function:

```python
    # Sampling shapes should support a variety of inputs
    stacked_means = torch.stack([M.mu0] * 5)
    s1 = M.sample(100)
    assert s1.shape == (100, M.ambient_dim), "Sampled points should have the correct shape"
    s2 = M.sample(100, z_mean=M.mu0)
    assert s2.shape == (100, M.ambient_dim), "Sampled points should have the correct shape"
    s3 = M.sample(z_mean=stacked_means)
    assert s3.shape == (5, M.ambient_dim), "Sampled points should have the correct shape"
    s3 = M.sample(100, z_mean=stacked_means)
    assert s3.shape == (500, M.ambient_dim), "Sampled points should have the correct shape"
```

Try setting seeds equivalently and verifying that this behaves as intended

**Comment by pchlenski** [comment]
@richarddushime thanks for all of your work. I have added comments for this PR, please review if you have time. Let me know if you have any questions or if anything is unclear.

Additionally a have a few general questions/comments:
- No random seeds are set. This might be OK for tests that are inside the call to `_shared_tests`, since we set the torch random seed prior to calling it, but for any new test functions involving sampling, please make sure to set a seed.
- Please change imports from `math` library to use their torch equivalents, e.g. `torch.pi`.
- In the future, please run the tests and verify that they pass. I can tell you didn't do this because, for instance, `Manifold` ojects have no `P` attribute.
- I'm not convinced that the extreme/edge case tests at the end of this PR will pass. It's a nice idea, but working with these manifolds can be pretty numerically unstable. A more moderate approach may be to carve out a range of curvatures, dimensionalities, etc., which are "officially supported," and have the tests reflect the extremes of that range. Let's talk more about this once you've addressed the rest of the comments / determined which of the final tests fail.

**Comment by richarddushime** [comment]
@pchlenski I have made another commit  to fix the tests and i m adding a clear description above

**Comment by richarddushime** [comment]
### below is general description of  whats covered and i tried to follow all the guidelines 
All tests covering the new changes pass; the remaining failures are ourside this PR scope
but i can fix them in a separate PR 

- **`3_verify_shapes.ipynb`**
  - Added general checks in `tests/test_manifolds.py::_shared_tests` that:
    - `dist2(x, y) ≈ dist(x, y) ** 2` and `sqrt(dist2(x, y)) ≈ dist(x, y)` for both `dist`/`pdist` on all manifolds.
  - Added `_product_manifold_tests` to verify for `ProductManifold`:
    - `pm.pdist2(X)` equals the sum of component `pdist2`s.
    - `pm.dist2(X1, X2)` equals the sum of component `dist2`s.

- **`2_polblogs_benchmark.ipynb`**
  - In `_product_manifold_tests`, added a check that `dist_component_by_manifold(pm, X1)` returns contributions that sum to 1.
  - In `tests/test_predictors.py`, added:
    - **`test_product_dt_matches_sklearn_on_euclidean`**: compares `ProductSpaceDT` vs sklearn `DecisionTreeClassifier` on a purely Euclidean `ProductManifold(signature=[(0.0, 4)])` via `manify.utils.benchmarks.benchmark`, asserting similar accuracies.
    - **`test_product_dt_matches_sklearn_regression_on_euclidean`**: analogous regression test vs sklearn `DecisionTreeRegressor`, asserting both models achieve small MSE on the same Euclidean problem.

- **`5_sampling.ipynb`**
  - **`test_sampling_distances_to_origin`**: checks that for a range of curvatures, distances from samples to `mu0` have reasonable mean/variance and are bounded on spherical manifolds.
  - **`test_sampling_consistency`**: verifies seeded equivalence of different sampling API variants (`n_samples` vs stacked `z_mean`, with/without `sigma`) across several curvatures.
  - **`test_sampling_edge_cases`**: exercises moderate curvature and covariance “edge” regimes to ensure sampling remains stable and points stay on the manifold.
  - **`_test_log_likelihood_properties`**: for both `Manifold` and `ProductManifold`, checks that log-likelihood differences between two parameterizations are finite and have non‑zero variance, and that the mean difference (approximate KL) is finite.

**Comment by pchlenski** [comment]
Hi Richard, thanks you for your new contribution. I've kicked off the testing workflow and will take a look at this PR over the weekend.
